src/datatable/datatable.jl

```diff
@@ -74,117 +74,105 @@ type DataTable <: AbstractDataTable
     end
-    return result
+    colnames = [Symbol(k) for (k,v) in kwargs]
+    columns = Any[v for (k,v) in kwargs]
+    DataTable(columns, Index(colnames))
 end
 
 function DataTable(columns::AbstractVector,
                    cnames::AbstractVector{Symbol} = gennames(length(columns)))
     return DataTable(convert(Vector{Any}, columns), Index(convert(Vector{Symbol}, cnames)))
 end
 
-
-# Initialize empty DataTable objects of arbitrary size
-function DataTable(t::Type, nrows::Integer, ncols::Integer)
-    columns = Vector{Any}(ncols)
-    for i in 1:ncols
-        columns[i] = NullableArray(t, nrows)
-    end
-    cnames = gennames(ncols)
-    return DataTable(columns, Index(cnames))
-end
-
 # Initialize an empty DataTable with specific eltypes and names
-function DataTable(column_eltypes::Vector, cnames::Vector, nrows::Integer)
-    p = length(column_eltypes)
-    columns = Vector{Any}(p)
-    for j in 1:p
-        columns[j] = NullableArray(column_eltypes[j], nrows)
+function DataTable{T<:Type}(column_eltypes::AbstractVector{T}, cnames::AbstractVector{Symbol}, nrows::Integer)
+    numcols = length(column_eltypes)
+    columns = Vector{Any}(numcols)
+    for j in 1:numcols
+        elty = column_eltypes[j]
+        if NullableArrays._isnullable(elty)
+            if eltype(elty) <: CategoricalValue
+                columns[j] = NullableCategoricalArray{eltype(elty)}(nrows)
+            else
+                columns[j] = NullableVector{eltype(elty)}(nrows)
+            end
+        else
+            if elty <: CategoricalValue
+                columns[j] = CategoricalVector{elty}(nrows)
+            else
+                columns[j] = Vector{elty}(nrows)
+            end
+        end
     end
-    return DataTable(columns, Index(cnames))
+    return DataTable(columns, Index(convert(Vector{Symbol}, cnames)))
 end
+
 # Initialize an empty DataTable with specific eltypes and names
 # and whether a nominal array should be created
-function DataTable(column_eltypes::Vector{DataType}, cnames::Vector{Symbol},
-                   nominal::Vector{Bool}, nrows::Integer)
-    p = length(column_eltypes)
-    columns = Vector{Any}(p)
-    for j in 1:p
-      if nominal[j]
-        columns[j] = NullableCategoricalArray{column_eltypes[j]}(nrows)
-      else
-        columns[j] = NullableArray{column_eltypes[j]}(nrows)
-      end
-    end
-    return DataTable(columns, Index(cnames))
-end
-
-# Initialize an empty DataTable with specific eltypes
-function DataTable(column_eltypes::Vector, nrows::Integer)
-    p = length(column_eltypes)
-    columns = Vector{Any}(p)
-    cnames = gennames(p)
-    for j in 1:p
-        columns[j] = NullableArray{column_eltypes[j]}(nrows)
+function DataTable{T<:Type}(column_eltypes::AbstractVector{T}, cnames::AbstractVector{Symbol},
+                            nominal::Vector{Bool}, nrows::Integer)
+    # Vector{DataType} -> Vector{Type}
+    updated_types = convert(Vector{Type}, column_eltypes)
+    for i in find(nominal)
```

Comment:
While this is not a big issue here, this allocates a new vector, so in general better loop over all items and have `nominal[i] || continue` at the beginning of the block.